pkg/eval/vals/aliased_types.go

```diff
@@ -7,8 +7,37 @@ import (
 	"github.com/xiaq/persistent/vector"
 )
 
-// File is an alias for *os.File.
-type File = *os.File
+// File wraps *os.File so that Elvish can treat the object as a read-only psuedo-map.
+type File struct {
+	OsFile *os.File `elvish:"hidden"`
+	// IoMode causes `$f[io-mode]` to return "ro" if the file is read-only, "rw" if it is
+	// read-write, and "wo" if it is write-only.
+	IoMode string
+}
+
+// IsStructMap tags File as a psuedo-map.
+func (File) IsStructMap() {}
```

Comment:
What is publically documented as pseudo-map in fact corresponds to two different internal concepts:

* A structmap is a pure data type - it is a struct that has a one-to-one correspondance between its struct fields and the fields exposed to Elvish

* A psuedo-structmap is any type that may be introspected like a map, but may contain other internal information.

The File type here is a pseudo-structmap and should be implemented as such (see https://github.com/elves/elvish/blob/f5aaba06911ac9d47d40adcb1f2fe615c77c741e/pkg/eval/vals/struct_map.go#L38)

Reply:
@xiaq: I'm still struggling, after several days of looking at the code, to see the point of the `PseudoStructMap` interface being a distinct type from `StructMap`. What I opted to do was add an `elvish:"hidden"` struct tag to flag that an otherwise public struct member should not be exposed by the Elvish pseudo-map mechanism. That avoids the extra layer of indirection incurred by the `PseudoStructMap` interface. It is also, IMHO, considerably simpler when compared to the `PseudoStructMap` interface.

Granted, I did that because I didn't understand the `PseudoStructMap` interface. I simply looked at how the `StructMap` interface worked and built on that by introducing support for the `elvish:"hidden"` struct tag. Nonetheless, I think my approach is preferable. I am confident that I can rewrite the existing uses of `PseudoStructMap` to instead use just the `StructMap` interface and greatly simplify the existing code.

Reply:
Struct maps are transparent; they are fully described by their fields - like maps, just with an immutable key set and type-constrained values. Pseudo-struct maps are opaque, but just happen to support introspection by the indexing operation (hence the "pseudo", since they are not really maps). This is why struct maps should not support hiding some of their fields.

The distinction between transparent and opaque types is not consequential for Elvish right now, but it will be in future when Elvish's type system get developed more. Many languages in the Lisp tradition have always made such a distinction - Racket has [opaque and transparent structs](https://docs.racket-lang.org/guide/define-struct.html#%28part._trans-struct%29) and [classes](https://docs.racket-lang.org/guide/classes.html), Clojure has [deftype and defrecord](https://clojure.org/reference/datatypes). The [Why have both deftype and defrecord?](https://clojure.org/reference/datatypes#_why_have_both_deftype_and_defrecord) section from the Clojure doc is especially insightful.

Even Java is catching up - although Java's OO style has always been geared towards opaque types, there's been a [recent proposal](https://cr.openjdk.java.net/~briangoetz/amber/datum.html) on introducing "data classes", something Google implemented as [@AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md) long ago.